Source/ASRunLoopQueue.mm

```diff
@@ -469,3 +489,230 @@ - (void)unlock
+  std::vector<id> itemsToProcess;
+
+  BOOL isQueueDrained = NO;
+  {
+    ASDN::MutexLocker l(_internalQueueLock);
+
+    NSInteger internalQueueCount = _internalQueue.count;
+    // Early-exit if the queue is empty.
+    if (internalQueueCount == 0) {
+      return;
+    }
+
+    ASSignpostStart(ASSignpostRunLoopQueueBatch);
+
+    /**
+     * For each item in the next batch, if it's non-nil then NULL it out
+     * and if we have an execution block then add it in.
+     * This could be written a bunch of different ways but
+     * this particular one nicely balances readability, safety, and efficiency.
+     */
+    NSInteger foundItemCount = 0;
+    for (NSInteger i = 0; i < internalQueueCount && foundItemCount < internalQueueCount; i++) {
+      /**
+       * It is safe to use unsafe_unretained here. If the queue is weak, the
+       * object will be added to the autorelease pool. If the queue is strong,
+       * it will retain the object until we transfer it (retain it) in itemsToProcess.
+       */
+      __unsafe_unretained id ptr = (__bridge id)[_internalQueue pointerAtIndex:i];
+      if (ptr != nil) {
+        foundItemCount++;
+        itemsToProcess.push_back(ptr);
+        [_internalQueue replacePointerAtIndex:i withPointer:NULL];
+      }
+    }
+
+    if (foundItemCount == 0) {
+      // If _internalQueue holds weak references, and all of them just become NULL, then the array
+      // is never marked as needsCompletion, and compact will return early, not removing the NULL's.
+      // Inserting a NULL here ensures the compaction will take place.
+      // See http://www.openradar.me/15396578 and https://stackoverflow.com/a/40274426/1136669
+      [_internalQueue addPointer:NULL];
+    }
+
+    [_internalQueue compact];
+    if (_internalQueue.count == 0) {
+      isQueueDrained = YES;
+    }
+  }
+
+  // itemsToProcess will be empty if _queueConsumer == nil so no need to check again.
+  auto count = itemsToProcess.size();
+  if (count > 0) {
+    as_activity_scope_verbose(as_activity_create("Process run loop queue batch", _rootActivity, OS_ACTIVITY_FLAG_DEFAULT));
+    auto itemsEnd = itemsToProcess.cend();
+    for (auto iterator = itemsToProcess.begin(); iterator < itemsEnd; iterator++) {
+      __unsafe_unretained id value = *iterator;
+      [value prepareForCATransactionCommit];
+      as_log_verbose(ASDisplayLog(), "processed %@", value);
+    }
+    if (count > 1) {
+      as_log_verbose(ASDisplayLog(), "processed %lu items", (unsigned long)count);
+    }
+  }
+
+  // If the queue is not fully drained yet force another run loop to process next batch of items
+  if (!isQueueDrained) {
+    CFRunLoopSourceSignal(_runLoopSource);
+    CFRunLoopWakeUp(_runLoop);
+  }
+
+  ASSignpostEnd(ASSignpostRunLoopQueueBatch);
+}
+
+- (void)enqueue:(id<ASCATransactionQueueObserving>)object
+{
+  if (!object) {
+    return;
+  }
+
+  if (_disableInterfaceStateCoalesce == YES) {
```

Comment:
Nit: remove `== YES`, I know it's used in other places but we can stop the madness.